json/circe/src/main/scala/sttp/client/circe/SttpCirceApi.scala

```diff
@@ -20,9 +21,27 @@ trait SttpCirceApi {
     * - `Left(HttpError(String))` if the response code was other than 2xx (deserialization is not attempted)
     * - `Left(DeserializationError)` if there's an error during deserialization
     */
-  def asJson[B: Decoder: IsOption]: ResponseAs[Either[ResponseError[io.circe.Error], B], Nothing] =
+  def asJson[B: Decoder: IsOption]: ResponseAs[Either[ResponseError[String,io.circe.Error], B], Nothing] =
     asString.map(ResponseAs.deserializeRightWithError(deserializeJson))
 
+  /**
+    * If the response is successful (2xx) or there the decoder is defined for given http code, tries to deserialize the body from a string into JSON. Returns:
+    * - `Right(b)` if the parsing was successful
+    * - `Left(HttpError(b))` if the response code was other than 2xx, but a decoder was defined
+    * - `Left(UnknownHttpError(String))` if the response code was other than 2xx and there was no decoder (deserialization is not attempted)
+    * - `Left(DeserializationError)` if there's an error during deserialization
+    */
+  def asJsonWithError[E:IsOption,B: Decoder: IsOption](knownErrorCodes:PartialFunction[StatusCode,Decoder[E]]): ResponseAs[Either[ResponseError[E,io.circe.Error], B], Nothing] = {
```

Comment:
Another option would be to re-use existing combinators:

```
def asJsonTypedErrors[T, E, D](
  success: ResponseAs[Either[ResponseError[String, D], T])( 
  errors: PartialFunction[StatusCode, ResponseAs[Either[ResponseError[String, D], E]]]): ResponseAs[Either[ResponseError[E, D], T]] = 
  fromMetadata { meta =>
    if (meta.isSuccess) success.map(Right(_)) else (...)
  }
```

which can be later used sth like:

```
val myResponse = asJsonTypedErrors(asJson[Success]) {
  case StatusCode.BadRequest => asJson[MyBadRequest]
  case StatusCode.Forbidden => asJson[MyForbidden]
}
```

One thing I don't really like is that the signatures become really complicated. I suspect most of the time you'd want errors to be represented as `String`s. Maybe we could use some type aliases here? E.g.:

```
abstract class ResponseErrorTyped[E, D] extends Exception

type ResponseError[E] = ResponseErrorTyped[E, String]
```

and maybe even json-lib-specific aliases, e.g. `type CirceResponseError[D] = ResponseError[io.circe.Error, D]`

Reply:
In fact this way we create a description which can be applied not only to json, so instead of `asJsonTypedErrors` we would have `asTypedErrors`.

And this can come in two variants:

```
asTypedErrors: ResponseAs[Either[ResponseError[E, D], T]] // the most general
asTypedErrorsUnsafe: ResponseAs[Either[E, T]]
```